jwst/dark_current/tests/test_dark_sub.py

```diff
@@ -0,0 +1,517 @@
+
+    np.testing.assert_array_equal(outfile.data[0, :, :, :], diff,
+                                  err_msg='dark file should be subtracted from sci file ')
+    np.testing.assert_array_equal(outfile.data[1, :, :, :], diff_int2,
+                                  err_msg='dark file should be subtracted from sci file ')
+
+
+def test_dark_skipped():
+    '''Verify that when the dark is not applied, the data is correctly flagged as such.'''
+
+    # size of integration
+    nints = 1
+    ngroups = 30
+    xsize = 1032
+    ysize = 1024
+
+    # create raw input data for step
+    dm_ramp = make_rampmodel(nints, ngroups, ysize, xsize)
+    dm_ramp.meta.exposure.nframes = 1
+    dm_ramp.meta.exposure.groupgap = 0
+
+    # populate data array of science cube
+    for i in range(0, ngroups-1):
+        dm_ramp.data[0, i, :, :] = i
+
+    refgroups = 20
+    # create dark reference file model with fewer frames than science data
+    dark = make_darkmodel(refgroups, ysize, xsize)
+
+    # populate data array of reference file
+    for i in range(0, refgroups - 1):
+        dark.data[0, i, :, :] = i * 0.1
+
+    # apply correction
+    outfile = darkcorr(dm_ramp, dark)
+
+    # get dark correction status from header
+    darkstatus = outfile.meta.cal_step.dark_sub
+    print('Dark status', darkstatus)
+
+    assert darkstatus == 'SKIPPED'
+
+
+def test_frame_avg():
+    '''Check that if NFRAMES>1 or GROUPGAP>0, the frame-averaged dark data are
+    subtracted group-by-group from science data groups and the ERR arrays are not modified'''
+
+    # size of integration
+    nints = 1
+    ngroups = 5
+    xsize = 1032
+    ysize = 1024
+
+    # create raw input data for step
+    dm_ramp = make_rampmodel(nints, ngroups, ysize, xsize)
+    dm_ramp.meta.exposure.nframes = 4
+    dm_ramp.meta.exposure.groupgap = 0
+
+    # populate data array of science cube
+    for i in range(0, ngroups-1):
+        dm_ramp.data[:, i, :, :] = i + 1
+
+    refgroups = 20
+    # create dark reference file model
+    dark = make_darkmodel(refgroups, ysize, xsize)
+
+    # populate data array of reference file
+    for i in range(0, refgroups - 1):
+        dark.data[0, i, :, :] = i * 0.1
+
+    # apply correction
+    outfile = darkcorr(dm_ramp, dark)
+
+    # dark frames should be averaged in groups of 4 frames
+    # this will result in average values of 0.15, 0.55, .095. 1.35
+
+    assert outfile.data[0, 0, 500, 500] == pytest.approx(0.85)
+    assert outfile.data[0, 1, 500, 500] == pytest.approx(1.45)
+    assert outfile.data[0, 2, 500, 500] == pytest.approx(2.05)
+    assert outfile.data[0, 3, 500, 500] == pytest.approx(2.65)
+
+    # check that the error array is not modified.
+    np.testing.assert_array_equal(outfile.err[:, :], 0,
+                                  err_msg='error array should remain 0 ')
+
+
+def make_rampmodel(nints, ngroups, ysize, xsize):
```

Comment:
It would be nice if these `make_rampmodel` and `make_darkmodel` functions used a consistent fixture format like that of `setup_nrc_cube` below.

Reply:
I'd like more flexibility in how I set up the sizes of the science and dark data arrays, so want to keep them as separate functions.